default/python/AI/ResourcesAI.py

```diff
@@ -148,115 +187,338 @@ def check_avoid_change_due_to_bureaucracy(self, pinfo: PlanetFocusInfo, focus: s
+    def calculate_planet_infos(self):  # noqa complexity
         """
-        # TODO this function depends on the specific rule that off-focus meter value are always the minimum value
+        Calculates for each possible focus an estimated target output of each planet and stores it in planet info.
+        """
+
+        def set_focus(focus_):
+            if focus_ in pinfo.planet.availableFoci and pinfo.planet.focus != focus_:
+                fo.issueChangeFocusOrder(pid, focus_)
+
         universe = fo.getUniverse()
-        unbaked_pids = [pid for pid in pids if pid not in self.baked_planet_info]
-        planet_infos = [(pid, self.all_planet_info[pid], self.all_planet_info[pid].planet) for pid in unbaked_pids]
-        for pid, pinfo, planet in planet_infos:
-            if INDUSTRY in planet.availableFoci and planet.focus != INDUSTRY:
-                fo.issueChangeFocusOrder(pid, INDUSTRY)  # may not be able to take, but try
-
-        def merge(po1, po2):
-            """merge two partially determined possible_outputs"""
-            return (
-                po1[0] or po2[0],
-                po1[1] or po2[1],
-                po1[2] or po2[2],
+        for pid, pinfo in self.planet_info.items():
+            set_focus(INDUSTRY)
+        universe.updateMeterEstimates(self.planet_ids)
+        for pid, pinfo in self.planet_info.items():
+            pinfo.possible_output[INDUSTRY] = self.possible_output_with_focus(pinfo.planet, INDUSTRY)
+            set_focus(RESEARCH)
+        universe.updateMeterEstimates(self.planet_ids)
+        for pid, pinfo in self.planet_info.items():
+            pinfo.possible_output[RESEARCH] = self.possible_output_with_focus(pinfo.planet, RESEARCH)
+            set_focus(INFLUENCE)
+        universe.updateMeterEstimates(self.planet_ids)
+        for pid, pinfo in self.planet_info.items():
+            pinfo.possible_output[INFLUENCE] = self.possible_output_with_focus(pinfo.planet, INFLUENCE)
+            set_focus(PROTECTION)
+        universe.updateMeterEstimates(self.planet_ids)
+        for pid, pinfo in self.planet_info.items():
+            pinfo.possible_output[PROTECTION] = self.possible_output_with_focus(pinfo.planet, PROTECTION)
+            # PROTECTION and GROWTH only produce through focus-less bonuses, so we can do without another meterEstimate
+            stability = PlanetUtilsAI.stability_with_focus(pinfo.planet, GROWTH)
+            pinfo.possible_output[GROWTH] = self.possible_output_with_focus(pinfo.planet, PROTECTION, stability)
+            # revert to current focus.
+            set_focus(pinfo.current_focus)
+        universe.updateMeterEstimates(self.planet_ids)
+        for pinfo in self.planet_info.values():
+            rated = sorted([(pinfo.possible_output[focus].rating, focus) for focus in supported_foci], reverse=True)
+            pinfo.best_output_focus = rated[0][1]
+            pinfo.second_best_focus = rated[1][1]
+        for pid, pinfo in self.planet_info.items():
```

Comment:
Maybe use a table?

Reply:
What do you mean by using a table?
Something like:
```
pinfo.rated_foci = sorted([(pinfo.possible_output[focus].rating, focus) for focus in supported_foci], reverse=True)
```

Reply:
I mean class Table from print utils

https://github.com/freeorion/freeorion/blob/master/default/python/tests/common/test_print_utils.py

Reply:
Good idea...